sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -186,98 +186,106 @@ case class CaseWhen(
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    // This variable holds the state of the result:
-    // -1 means the condition is not met yet and the result is unknown.
-    val NOT_MATCHED = -1
-    // 0 means the condition is met and result is not null.
-    val HAS_NONNULL = 0
-    // 1 means the condition is met and result is null.
-    val HAS_NULL = 1
-    // It is initialized to `NOT_MATCHED`, and if it's set to `HAS_NULL` or `HAS_NONNULL`,
-    // We won't go on anymore on the computation.
-    val resultState = ctx.freshName("caseWhenResultState")
-    ev.value = JavaCode.global(
-      ctx.addMutableState(CodeGenerator.javaType(dataType), ev.value),
-      dataType)
+    if (branches.length == 1) {
+      // If we have only single branch we can use If expression and its codeGen
+      If(
+        branches(0)._1,
+        branches(0)._2,
+        elseValue.getOrElse(Literal.create(null, branches(0)._2.dataType))).doGenCode(ctx, ev)
+    } else {
+      // This variable holds the state of the result:
```

Comment:
we can create a method `multiBranchesCodegen` to put the original code:
```
override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
  if ...
  else {
    multiBranchesCodegen
  }
}
```
then the diff would be much smaller.

Reply:
yes, it makes sense